src/main/java/org/apache/sysml/hops/estim/EstimatorMatrixHistogram.java

```diff
@@ -71,15 +71,66 @@ public double estim(MMNode root) {
 		
 		return ret;
 	}
-
-	@Override
+	
+	@Override 
 	public double estim(MatrixBlock m1, MatrixBlock m2) {
 		MatrixHistogram h1 = new MatrixHistogram(m1, _useExcepts);
 		MatrixHistogram h2 = (m1 == m2) ? //self product
 			h1 : new MatrixHistogram(m2, _useExcepts);
 		return estimIntern(h1, h2);
 	}
 
+	public double estim(MatrixBlock m1, MatrixBlock m2, String op) {
+		MatrixHistogram h1 = new MatrixHistogram(m1, _useExcepts);
+		MatrixHistogram h2 = (m1 == m2) ? //self product
+			h1 : new MatrixHistogram(m2, _useExcepts);
+		double r_sparsity = 0;
+		double c_sparsity = 0;
+		double msize = m1.getNumRows()*m1.getNumColumns();
+		switch (op) {
+		case "mult":
+			for(int k=0; k<h1.getCols();k++) {
+				c_sparsity = c_sparsity + ((double)h1.cNnz[k]/msize)*((double)h2.cNnz[k]/msize);
+			}
+			for(int j=0; j<h1.getRows();j++) {
+				r_sparsity = r_sparsity + ((double)h1.rNnz[j]/msize)*((double)h2.rNnz[j]/msize);
+			}
+			return Math.min(c_sparsity, r_sparsity);
+		case "plus":
+			for(int k=0; k<h1.getCols();k++) {
+				c_sparsity = c_sparsity + (double)h1.cNnz[k]/msize + (double)h2.cNnz[k]/msize - ((double)h1.cNnz[k]/msize)*((double)h2.cNnz[k]/msize);
+			}
+			for(int j=0; j<h1.getRows();j++) {
+				r_sparsity = r_sparsity + (double)h1.rNnz[j]/msize + (double)h2.rNnz[j]/msize - ((double)h1.rNnz[j]/msize)*((double)h2.rNnz[j]/msize);
+			}
+			return Math.min(c_sparsity, r_sparsity);
+		case "cbind":
+			return (double) m1.getNonZeros()+m2.getNonZeros()/(m1.getNumRows()*(m1.getNumColumns()+1));
+		case "rbind":
+			return (double) m1.getNonZeros()+m2.getNonZeros()/((m1.getNumRows()+1)*m1.getNumColumns());
```

Comment:
The operations cbind and rbind may concatenate arbitrary matrices - hence, we need to take the exact number of columns and rows form m2 into account to compute the sparsity.

Reply:
Also try to use OptimizerUtils.getSparsity() to prevent int overflows when doing rows*cols.